pkg/service/application_metadata.go

```diff
@@ -0,0 +1,63 @@
+package service
+
+import (
+	"context"
+	"errors"
+
+	"github.com/pyroscope-io/pyroscope/pkg/model"
+	"github.com/pyroscope-io/pyroscope/pkg/storage"
+	"gorm.io/gorm"
+)
+
+type ApplicationMetadataService struct {
+	db *gorm.DB
+}
+
+func NewApplicationMetadataService(db *gorm.DB) ApplicationMetadataService {
+	return ApplicationMetadataService{db: db}
+}
+
+func (svc ApplicationMetadataService) List(ctx context.Context) (apps []storage.ApplicationMetadata, err error) {
+	tx := svc.db.WithContext(ctx)
+	result := tx.Find(&apps)
+	return apps, result.Error
+}
+
+func (svc ApplicationMetadataService) Get(ctx context.Context, name string) (storage.ApplicationMetadata, error) {
+	app := storage.ApplicationMetadata{}
+	if err := model.ValidateAppName(name); err != nil {
+		return app, err
+	}
+
+	tx := svc.db.WithContext(ctx)
+	res := tx.Where("fq_name = ?", name).First(&app)
+
+	switch {
+	case errors.Is(res.Error, gorm.ErrRecordNotFound):
+		return app, model.ErrApplicationNotFound
+	default:
+		return app, res.Error
+	}
+}
+
+func (svc ApplicationMetadataService) CreateOrUpdate(ctx context.Context, application storage.ApplicationMetadata) error {
+	if err := model.ValidateAppName(application.FQName); err != nil {
```

Comment:
I think if the app name made it all the way here is a bit too late to do validations. I think they should either happen much earlier (in the controller), or not happen at all.

Reply:
Well that's the approach we use in other places ¯\\\_(ツ)\_\/¯

https://github.com/pyroscope-io/pyroscope/blob/34edfddebbbd33c6dbdb7925c5d95fddef8f5200/pkg/service/user.go#L18
https://github.com/pyroscope-io/pyroscope/blob/34edfddebbbd33c6dbdb7925c5d95fddef8f5200/pkg/service/api_key.go#L17
https://github.com/pyroscope-io/pyroscope/blob/34edfddebbbd33c6dbdb7925c5d95fddef8f5200/pkg/service/annotation.go#L23

Reply:
Also, I can remember one issue when invalid (empty) app names were generating due to some flaws in parsing. It was hard to fix. This validation would allow us to identify the error quickly.

Reply:
Fair enough, let's validate everywhere then :)